src/protocol/handshake.rs

```diff
@@ -145,28 +237,46 @@ impl Handshake for InnerNode {
                 let sig = create_session_signature(&self.crypto, &shared_value);
 
                 // prepare the HTTP request message
-                let mut request = Vec::new();
-                request.extend_from_slice(b"GET / HTTP/1.1\r\n");
-                request.extend_from_slice(format!("User-Agent: {}\r\n", hs_cfg.ident).as_bytes());
-                request.extend_from_slice(b"Connection: Upgrade\r\n");
-                request.extend_from_slice(b"Upgrade: XRPL/2.0, XRPL/2.1, XRPL/2.2\r\n"); // TODO: which ones should we handle?
-                request.extend_from_slice(b"Connect-As: Peer\r\n");
-                request.extend_from_slice(format!("Public-Key: {base58_pk}\r\n").as_bytes());
-                request.extend_from_slice(format!("Session-Signature: {sig}\r\n").as_bytes());
-                // txrr - enables transaction relay
-                // ledgerreplay - enables ledger replay
-                request.extend_from_slice(b"X-Protocol-Ctl: txrr=1;ledgerreplay=1\r\n");
-                request.extend_from_slice(b"\r\n");
-                let request = Bytes::from(request);
+                let mut req = Vec::new();
+                let mut req_header = |mut header: String| {
+                    // Append `\r\n' to every header.
+                    header.push_str("\r\n");
+                    req.extend_from_slice(header.as_bytes());
+                };
+
+                req_header("GET / HTTP/1.1".into());
+                req_header(format!("User-Agent: {}", hs_cfg.http_ident));
+                req_header(format!("Upgrade: {}", hs_cfg.http_upgrade_req));
+                req_header(format!("Connection: {}", hs_cfg.http_connection));
+                req_header(format!("Connect-As: {}", hs_cfg.http_connect_as));
+                if let Some(ref crawl) = hs_cfg.http_crawl {
+                    req_header(format!("Crawl: {crawl}"))
+                };
+                req_header(format!("X-Protocol-Ctl: {}", hs_cfg.http_x_protocol_ctl));
+                if let Some(ref time) = hs_cfg.http_network_time {
+                    req_header(format!("Network-Time: {time}"))
+                };
+                req_header(format!("Public-Key: {base58_pk}"));
+                req_header(format!("Session-Signature: {sig}"));
+                if let Some(ref ledger) = hs_cfg.http_closed_ledger {
+                    req_header(format!("Closed-Ledger: {ledger}"))
+                };
+                if let Some(ref ledger) = hs_cfg.http_prev_ledger {
+                    req_header(format!("Previous-Ledger: {ledger}"))
+                };
+                if let Some(ref header) = hs_cfg.http_unexpected_extra_field_and_value {
+                    req_header(header.clone())
+                };
+                req_header("".into()); // An HTTP header ends with '\r\n'
```

Comment:
```suggestion
                req_header("".into()); // A HTTP header ends with '\r\n'
```

Reply:
let's keep it this [way](https://www.pristineword.com/grammar-an-http/)